docs: technical guidelines for contributing (#8562)
elastic/eui: wiki/contributing-to-eui/documenting/new.md

The folder structure matches the navigation and content structure, similar to common static site builders like Jekyll. For instance, for editing the Components / Containers / Accordion page, the content will be found in the `docs/components/containers/accordion.mdx` file.

> [!TIP]
> For formatting content always prefer using Markdown over React

### Document structure

All component pages should follow the following architecture:

- Intro
- Component: a bare-bones example of the component in its default form
- Usage (optional): examples of different states and use-cases
- Guidelines (optional): dos and donts
- Props

### Partials

A document can be split into different files when content is too long, for easier editing.

Markdown files prefixed with an underscore will be ignored when generating the sidebar. These hidden files can then be imported and rendered in the main document.

For example, the document for the Modal page lives in `containers/modal/index.mdx`:

```
containers/
  modal/
    _guidelines.mdx

>     _patterns.mdx

Reviewer: I'm wondering, shouldn't we name the file the same as we name the section? Either both `Patterns` or `Usage`.

Author: yes, I took that from the actual real thing so wanted to stay true to it but it makes sense we change it to match 😀

    index.mdx
```

The Usage and Guidelines sections have their own file that can be imported:

```mdx
<!-- in index.mdx -->

## Usage

> import Patterns from './_patterns.mdx';

Author: will add `mdx-code-block`, good one 👍 

> if it's .mdx file that we're importing, is it okay? Does this only apply to .tsx files?

that's a good question, I think it should be OK, it seems to be part of MDX → https://mdxjs.com/docs/using-mdx/#mdx-content

> import Patterns from './_patterns.mdx';

Reviewer: Oh, I didn't mean importing MDX files, that's ofc part of the MDX API. What I meant is, should we wrap all imports with `mdx-code-block`, even `.mdx` files or only the code ones? Because we do want to view that MDX content we're importing directly or at least know that there's supposed to be MDX content there,`mdx-code-block` might hide it. Not sure how it's shown in the Markdown viewers, haven't checked 😄

> import Patterns from './_patterns.mdx';

Author: argh, sorry I misunderstood, the question was clear already — I would say no, `.mdx` imports I would leave without, you agree?

> import Patterns from './_patterns.mdx';

Reviewer: No worries 😄 I agree!

> import Patterns from './_patterns.mdx';

Author: fixed ✨

<Patterns />

## Guidelines

import Guidelines from './_guidelines.mdx';

<Guidelines />
```

Read more [about importing markdown](https://docusaurus.io/docs/markdown-features/react#importing-markdown) in the Docusaurus docs.

### Collapsible sidebar items

To make a folder a collapsible item in the sidebar, add a `_category_.yml` file that includes a boolean `collapsed` field. Usually set to `false`.

Other fields that can be useful are `label` and `position`.

> It's also possible to make a collapsible item also be a document. Set the `link` field to have `type: doc` and `id: <DOCUMENT_ID>`, with the `id` from the document it should open. If the `id` is not explicitly set in the document file's front matter, it's the relative path starting from the `docs` folder, without the file extension, e.g. `docs/components/containers/accordion`

Reviewer: ```suggestion
It's possible to make a collapsible item be a document as well. Set the `link` field to have `type: doc` and `id: <DOCUMENT_ID>`, with the `id` from the document it should open. If the `id` is not explicitly set in the document file's front matter, it's the relative path starting from the `docs` folder, without the file extension, e.g. `docs/components/containers/accordion`
```

Reviewer: Maybe we can also make a link for the `front matter`:

```suggestion
It's also possible to make a collapsible item also be a document. Set the `link` field to have `type: doc` and `id: <DOCUMENT_ID>`, with the `id` from the document it should open. If the `id` is not explicitly set in the document file's [front matter](#front-matter), it's the relative path starting from the `docs` folder, without the file extension, e.g. `docs/components/containers/accordion`
```

> [!WARNING]
> Nesting items in the sidebar should be avoided when possible, deeply nested items can make the menu harder to use

### Markdown

As previously mentioned, using Markdown is highly encouraged. It makes it easier for anybody to contribute to the documentation.

During the build step of the website, elements with a counterpart in EUI will be accordingly replaced, e.g. a link with `EuiLink`, a inline code block with `EuiCode`, etc.

#### Front matter
TODO
https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#markdown-front-matter

#### Headings

Headings in Markdown are created with a hash symbol in a new line. The number of hashes will determine the heading level, # for 1, ## for 2, etc.

> [!IMPORTANT]
> Do not use heading levels for styling, use them to properly structure the document hierarchically

#### Links

Markdown is the recommended syntax for creating links. Docusaurus will take care of rendering it correctly.

```md
[This is a link](https://example.com)
```

To link between documents, use **relative file paths, including the file extension**, e.g. `[EuiPanel](../containers/panel/index.mdx)`

> 

Reviewer: It could be useful to add the source here: https://docusaurus.io/docs/markdown-features/links

Sth like "`You can read more about it [here](https://docusaurus.io/docs/markdown-features/links)`".

#### Admonitions

The syntax for creating admonitions is wrapping a paragraph in three colons, e.g. `:::`. The type of admonition and title can be defined after the first three colons. This will render `EuiCallOut`.

```mdx
:::info This is the title

Here's some content

:::
```

> There are 5 types of admoniotions supported, with an icon and color respectively.

Reviewer: ```suggestion
There are 5 types of admonitions supported, with an icon and color respectively.
```

Are you using VSCode? There's a very useful extension - Code Spell Checker from Street Side Software. It underlines the typos in blue in the editor.

Author: I'm using Zed 😑 (such an ugly typo)

Reviewer: There seem to be a couple of options for Zed though: https://github.com/zed-industries/zed/issues/5305#issuecomment-2692429997

- note
- tip
- info
- accessibility
- warning
- danger

#### Code blocks, live demos

A code block or code snippet can be created by wrapping text in backticks, e.g. \`code\`.

For a simple one-liner, single backticks are enough. To create a multiline snippet or code block, use three backticks. For syntax highlighting, the language name can be added right after the first three backticks.

\```html  
<span>Hello World</span>  
\```

Code blocks can be used to generate live demos, where a React component that is exported in the snippet, will be rendered on the page. To do that, append the word "interactive" after the language name, usually `tsx`.

\```tsx interactive  
// TypeScript code goes here  
\```

> 

Reviewer: We should probably also mention the explicit usage of the `Demo` component and a bit about its properties.

Author: that's exactly it, will link 👍
